File: backend/app/api/v1/reports.py
"""Impact-report endpoints — aggregated vaccination metrics."""

from __future__ import annotations

import csv
import io
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_active_user
from app.models.coverage import CoverageFacility
from app.models.user import User

router = APIRouter(prefix="/reports", tags=["reports"])


# ── Schemas ──────────────────────────────────────────────────────────────────


class CoverageByCountry(BaseModel):
    country: str
    facilityCount: int
    avgCoverageRate: float
    totalDosesAdministered: int


class StockSummary(BaseModel):
    status: str
    facilityCount: int


class ColdChainSummary(BaseModel):
    totalReadings: int
    breachCount: int
    complianceRate: float
    totalAlerts: int
    resolvedAlerts: int


class WastageSummary(BaseModel):
    totalWastageQty: int
    totalIssuedQty: int
    wastageRate: float


class FacilityPerformance(BaseModel):
    id: str
    name: str
    region: str
    country: str
    vaccineType: str
    dosesAdministered: int
    targetPopulation: int
    coverageRate: float
    stockStatus: str
    current: int


class ImpactReportData(BaseModel):
    generatedAt: str
    dateFrom: str | None = None
    dateTo: str | None = None
    coverageByCountry: list[CoverageByCountry]
    stockSummary: list[StockSummary]
    coldChain: ColdChainSummary
    wastage: WastageSummary
    facilityPerformance: list[FacilityPerformance]


# ── Helpers ───────────────────────────────────────────────────────────────────


def _apply_filters(stmt, date_from: str | None, date_to: str | None, country: str | None):
    """Apply optional date / country filters to a CoverageFacility query."""
    if country:
        stmt = stmt.where(CoverageFacility.country == country)
    # date filters are applied only when the model exposes a period column
    return stmt


async def _build_report(
    db: AsyncSession,
    date_from: str | None = None,
    date_to: str | None = None,
    country: str | None = None,
) -> ImpactReportData:
    """Aggregate DB rows into the report payload."""

    base = select(CoverageFacility)
    base = _apply_filters(base, date_from, date_to, country)

    result = await db.execute(base.order_by(CoverageFacility.country, CoverageFacility.name))
    rows = result.scalars().all()

    # -- coverage by country --------------------------------------------------
    country_map: dict[str, dict] = {}
    for r in rows:
        entry = country_map.setdefault(r.country, {"count": 0, "cov_sum": 0.0, "doses": 0})
        entry["count"] += 1
        entry["cov_sum"] += r.coverage_rate
        entry["doses"] += r.doses_administered

    coverage_by_country = [
        CoverageByCountry(
            country=c,
            facilityCount=v["count"],
            avgCoverageRate=round(v["cov_sum"] / v["count"], 1) if v["count"] else 0,
            totalDosesAdministered=v["doses"],
        )
        for c, v in country_map.items()
    ]

    # -- stock summary --------------------------------------------------------
    status_map: dict[str, int] = {}
    for r in rows:
        status_map[r.stock_status] = status_map.get(r.stock_status, 0) + 1
    stock_summary = [StockSummary(status=s, facilityCount=n) for s, n in status_map.items()]

    # -- cold-chain (placeholder until cold_chain model is wired) -------------
    cold_chain = ColdChainSummary(
        totalReadings=0,
        breachCount=0,
        complianceRate=100.0,
        totalAlerts=0,
        resolvedAlerts=0,
    )

    # -- wastage (placeholder) ------------------------------------------------
    wastage = WastageSummary(totalWastageQty=0, totalIssuedQty=0, wastageRate=0.0)

    # -- facility performance -------------------------------------------------
    facility_perf = [
        FacilityPerformance(
            id=str(r.id),
            name=r.name,
            region=r.region,
            country=r.country,
            vaccineType=r.vaccine_type,
            dosesAdministered=r.doses_administered,
            targetPopulation=r.target_population,
            coverageRate=r.coverage_rate,
            stockStatus=r.stock_status,
            current=0,
        )
        for r in rows
    ]

    return ImpactReportData(
        generatedAt=datetime.now(timezone.utc).isoformat(),
        dateFrom=date_from,
        dateTo=date_to,
        coverageByCountry=coverage_by_country,
        stockSummary=stock_summary,
        coldChain=cold_chain,
        wastage=wastage,
        facilityPerformance=facility_perf,
    )


# ── Endpoints ─────────────────────────────────────────────────────────────────


@router.get("/impact", response_model=ImpactReportData, summary="Aggregated impact report")
async def get_impact_report(
    dateFrom: str | None = Query(default=None),
    dateTo: str | None = Query(default=None),
    country: str | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_active_user),
) -> ImpactReportData:
    """Return aggregated vaccination impact data across all facilities."""
    return await _build_report(db, date_from=dateFrom, date_to=dateTo, country=country)


@router.get("/impact/csv", summary="Impact report as CSV download")
async def get_impact_report_csv(
    dateFrom: str | None = Query(default=None),
    dateTo: str | None = Query(default=None),
    country: str | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_active_user),
):
    """Stream the facility-performance section of the impact report as CSV."""
    report = await _build_report(db, date_from=dateFrom, date_to=dateTo, country=country)

    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(["id", "name", "region", "country", "vaccineType",
                     "dosesAdministered", "targetPopulation", "coverageRate",
                     "stockStatus", "current"])
    for f in report.facilityPerformance:
        writer.writerow([f.id, f.name, f.region, f.country, f.vaccineType,
                         f.dosesAdministered, f.targetPopulation, f.coverageRate,
                         f.stockStatus, f.current])
    buf.seek(0)

    return StreamingResponse(
        iter([buf.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=impact_report.csv"},
    )

File: frontend/src/api/reports.ts
import { apiClient, isDemoMode } from "./client";

/* ── Types ─────────────────────────────────────────────── */

export interface CoverageByCountry {
  country: string;
  facilityCount: number;
  avgCoverageRate: number;
  totalDosesAdministered: number;
}

export interface StockSummary {
  status: string;
  facilityCount: number;
}

export interface ColdChainSummary {
  totalReadings: number;
  breachCount: number;
  complianceRate: number;
  totalAlerts: number;
  resolvedAlerts: number;
}

export interface WastageSummary {
  totalWastageQty: number;
  totalIssuedQty: number;
  wastageRate: number;
}

export interface FacilityPerformance {
  id: string;
  name: string;
  region: string;
  country: string;
  vaccineType: string;
  dosesAdministered: number;
  targetPopulation: number;
  coverageRate: number;
  stockStatus: string;
  current: number;
}

export interface ImpactReportData {
  generatedAt: string;
  dateFrom?: string;
  dateTo?: string;
  coverageByCountry: CoverageByCountry[];
  stockSummary: StockSummary[];
  coldChain: ColdChainSummary;
  wastage: WastageSummary;
  facilityPerformance: FacilityPerformance[];
}

export interface ReportParams {
  dateFrom?: string;
  dateTo?: string;
  country?: string;
}

/* ── Demo / mock data ──────────────────────────────────── */

const MOCK_REPORT: ImpactReportData = {
  generatedAt: new Date().toISOString(),
  dateFrom: "2025-01-01",
  dateTo: "2025-12-31",
  coverageByCountry: [
    { country: "Nigeria", facilityCount: 420, avgCoverageRate: 78, totalDosesAdministered: 1_240_000 },
    { country: "Kenya", facilityCount: 310, avgCoverageRate: 84, totalDosesAdministered: 680_000 },
    { country: "Ghana", facilityCount: 180, avgCoverageRate: 91, totalDosesAdministered: 320_000 },
    { country: "Senegal", facilityCount: 145, avgCoverageRate: 72, totalDosesAdministered: 190_000 },
  ],
  stockSummary: [
    { status: "Adequate", facilityCount: 780 },
    { status: "Low", facilityCount: 210 },
    { status: "Stockout", facilityCount: 65 },
  ],
  coldChain: {
    totalReadings: 48_000,
    breachCount: 320,
    complianceRate: 99,
    totalAlerts: 142,
    resolvedAlerts: 138,
  },
  wastage: {
    totalWastageQty: 24_500,
    totalIssuedQty: 2_430_000,
    wastageRate: 1.01,
  },
  facilityPerformance: [
    { id: "f1", name: "Lagos Hub", region: "Lagos", country: "Nigeria", vaccineType: "bOPV", dosesAdministered: 125_000, targetPopulation: 150_000, coverageRate: 83, stockStatus: "Adequate", current: 4200 },
    { id: "f2", name: "Kano Facility", region: "Kano", country: "Nigeria", vaccineType: "PENTA", dosesAdministered: 98_000, targetPopulation: 120_000, coverageRate: 82, stockStatus: "Low", current: 980 },
    { id: "f3", name: "Nairobi Store", region: "Nairobi", country: "Kenya", vaccineType: "PCV13", dosesAdministered: 67_000, targetPopulation: 80_000, coverageRate: 84, stockStatus: "Adequate", current: 3100 },
  ],
};

/* ── API functions ─────────────────────────────────────── */

export async function getImpactReport(params?: ReportParams): Promise<ImpactReportData> {
  if (isDemoMode()) return MOCK_REPORT;
  const { data } = await apiClient.get<ImpactReportData>("/reports/impact", { params });
  return data;
}

export function getImpactReportCsvUrl(params?: ReportParams): string {
  const base = apiClient.defaults.baseURL ?? "/api/v1";
  const qs = new URLSearchParams();
  if (params?.dateFrom) qs.set("dateFrom", params.dateFrom);
  if (params?.dateTo) qs.set("dateTo", params.dateTo);
  if (params?.country) qs.set("country", params.country);
  const query = qs.toString();
  return `${base}/reports/impact/csv${query ? `?${query}` : ""}`;
}
